src/orchestration/training_cluster_monitor.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2022 Google LLC
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     https://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Monitor training clusters.
+
+Restarts training cluster if detects training not progressing.
+"""
+import os
+import re
+import datetime
+import time
+import subprocess
+import sys
+
+from absl import app
+from absl import flags
+from absl import logging
+from google.cloud import storage
+
+FLAGS = flags.FLAGS
+
+
+
+flags.DEFINE_string('project_id', '', 'GCP project id.')
+flags.DEFINE_string('checkpoint_path', '',
+                    'GCS prefix to model checkpoint_file.')
+flags.DEFINE_integer('train_ttl_sec', 2400,
```

Comment:
Can we give the user an option to stop trying after a few restarts?

Reply:
This seems additional features. Can you please file a feature request and we can evaluate the priority?

Reply:
It's part of this feature. We don't want the restart to keep happening forever if there is no progress.

Reply:
Can you please list out the CUJ of this feature? There can be a lot of optimizations we can add in here.

Reply:
I think your solution covers all the CUJ. The only thing is that we don't want to keep retrying forever. A reasonable termination condition would if there is no progress after a few retries. Either this or a total limit on the retry would be acceptable.

Reply:
Done